pkg/controller/flow/reconcile.go

```diff
@@ -0,0 +1,360 @@
+	resourceClient, err := CreateResourceInterface(r.restConfig, ref, namespace)
+	if err != nil {
+		glog.Warningf("failed to create dynamic client resource: %v", err)
+		return "", err
+	}
+
+	obj, err := resourceClient.Get(ref.Name, metav1.GetOptions{})
+	if err != nil {
+		glog.Warningf("failed to get object: %v", err)
+		return "", err
+	}
+	status, ok := obj.Object["status"]
+	if !ok {
+		return "", fmt.Errorf("%q does not contain status", ref.Name)
+	}
+	statusMap := status.(map[string]interface{})
+	serviceName, ok := statusMap[targetFieldName]
+	if !ok {
+		return "", fmt.Errorf("%q does not contain field %q in status", targetFieldName, ref.Name)
+	}
+	serviceNameStr, ok := serviceName.(string)
+	if !ok {
+		return "", fmt.Errorf("%q status field %q is not a string", targetFieldName, ref.Name)
+	}
+	return serviceNameStr, nil
+}
+
+func (r *reconciler) reconcileChannel(flow *v1alpha1.Flow) (*channelsv1alpha1.Channel, error) {
+	channelName := flow.Name
+
+	channel := &channelsv1alpha1.Channel{}
+	err := r.client.Get(context.TODO(), client.ObjectKey{Namespace: flow.Namespace, Name: channelName}, channel)
+	if errors.IsNotFound(err) {
+		channel, err = r.createChannel(flow)
+		if err != nil {
+			glog.Errorf("Failed to create channel %q : %v", channelName, err)
+			return nil, err
+		}
+	} else if err != nil {
+		glog.Errorf("Failed to reconcile channel %q failed to get channels : %v", channelName, err)
+		return nil, err
+	}
+
+	// Should make sure channel is what it should be. For now, just assume it's fine
+	// if it exists.
+	return channel, err
+}
+
+func (r *reconciler) createChannel(flow *v1alpha1.Flow) (*channelsv1alpha1.Channel, error) {
+	channelName := flow.Name
+	channel := &channelsv1alpha1.Channel{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      channelName,
+			Namespace: flow.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*r.NewControllerRef(flow),
+			},
+		},
+		Spec: channelsv1alpha1.ChannelSpec{
+			ClusterBus: defaultBusName,
+		},
+	}
+	if err := r.client.Create(context.TODO(), channel); err != nil {
+		return nil, err
+	}
+	return channel, nil
+}
+
+func (r *reconciler) reconcileSubscription(channelName string, target string, flow *v1alpha1.Flow) (*channelsv1alpha1.Subscription, error) {
+	subscriptionName := flow.Name
+
+	subscription := &channelsv1alpha1.Subscription{}
+	err := r.client.Get(context.TODO(), client.ObjectKey{Namespace: flow.Namespace, Name: subscriptionName}, subscription)
+	if errors.IsNotFound(err) {
+		subscription, err = r.createSubscription(channelName, target, flow)
+		if err != nil {
+			glog.Errorf("Failed to create subscription %q : %v", subscriptionName, err)
+			return nil, err
+		}
+	} else if err != nil {
+		glog.Errorf("Failed to reconcile subscription %q failed to get subscriptions : %v", subscriptionName, err)
+		return nil, err
+	}
+
+	// Should make sure subscription is what it should be. For now, just assume it's fine
+	// if it exists.
+	return subscription, err
```

Comment:
Seems like `err` can only be `nil` here, so it might be clearer to `return channel, nil`.

Reply:
done